Canary blog post (#262)
strimzi/strimzi.github.io: _posts/2021-11-07-canary.md

---
layout: post

> title: "Use a canary in your Apache Kafka coal mine"

Reviewer: What about something like: `Why should you use a canary in your Apache Kafka coal mine?`

Should we capitalize `Canary` as it's the name of the component?

date: 2021-11-08
author: paolo_patierno
---

The Strimzi cluster operator is really great to deploy your Kafka cluster on Kubernetes and handling everything related to it.
It helps with brokers configuration, managing the security aspects, doing the needed upgrades, scaling up and down and many more.
But ... what if you want to have more insights about how the Kafka cluster is behaving?
What if you want to know if the Kafka brokers are operating correctly so that clients can exchange messages?

> This is where a new component in the Strimzi family comes into the picture ... please welcome to the canary!

Reviewer: I think we should have either:
`Welcome to the canary!`
or 
`Please welcome the canary`

But having both `please` and `to` sounds strange

<!--more-->

> ## How a canary can help in a coal mine?

Reviewer: -> `How can a canary help in a coal mine?`

> When a Kafka cluster is successfully deployed using the Strimzi operator, the next step is to use it with your own applications or maybe, depending on your business, allowing external users to use it even for integration with your solution.

Reviewer 2: Maybe _depending on your use-case_ might be better than _depending on your business_?

Having information about producing and consuming latency, as well as connection times, errors on exchanging messages and so on is really useful for that purpose.
If your Kafka cluster is exposed to external users you would like to know if they are able to use the cluster itself; the canary can be used to simulate activity for identifying problems from a user perspective even when the cluster seems to be operating correctly.
Another use case could be when you are handling a fleet of Kafka clusters that you want to update to a newer version doing a canary rolling; just updating a few of them, checking that everything is fine and only then continuing with the rest of the fleet.

These are the main problems that the canary tries to help with.
Of course, it is not a replacement for any kind of monitoring solution that can be used to show metrics coming from the Kafka cluster, as provided by the corresponding Strimzi examples for setting up Prometheus and Grafana together with useful dashboards.

> Like the coal miners using the canaries in the coal mines for early warning about toxic gases in the 20th century, the Strimzi canary can be used to raise alerts when the Kafka cluster is starting to behave badly.

Reviewer 2: I wonder if you can find some nice picture or meme to include here ;-). I also wonder if everyone knows the history behind it ... should we maybe add a link to https://en.wikipedia.org/wiki/Sentinel_species or some other web for those who might not understand it?

Author: Added the link for now, searching for a good picture or meme ;-)

> ## What the canary provides?

Reviewer: -> `What does the canary provide?`

The canary tool is made by several internal services, used for different purposes:

> * A `topic` service that is in charge to create and handle the dedicated topic (named `__strimzi_canary` by default) used by the canary producer and consumer to exchange messages.

Reviewer: `in charge of creating and handling`

* A `consumer` service that is used to receive the messages sent by the producer through the canary topic.
* A `connection` service that just checks the connection to the brokers periodically; it opens and then immediately close a connection to each broker just to validate that it's reachable.
* A `status` service that aggregates some of the producer and consumer metrics exposing them on a dedicated HTTP endpoint in a JSON format.

When the canary starts for the first time, the `topic` service creates the dedicated topic with a number of partitions equals to the number of brokers, so that for each broker there is one partition.
If the Kafka cluster is scaled up, the service adds more partitions to the topic assigning them in order to have the new partitions on the new brokers; the objective is always having one partition per broker so that when the producer and consumer exchange messages through the topic they are "testing" somehow all the brokers in the cluster using all partitions.

> If the Kafka cluster is scaled down, because it's not possible to remove partitions from a topic, the canary will just avoid using excess partitions (maybe reusing them if the cluster is scaled up again).

Reviewer: `maybe reusing them` so I'm wondering if it is doing it or not?

Author: Is your point about removing `maybe`, to be clear that it's going to reuse them when scale up again?

Reviewer: Yes. With `maybe` it kind of feels like it may or may not happen.

> While `producer` and `consumer` services exchange messages through the dedicated topic, different metrics are gathered and being exposed via the HTTP `/metrics` endpoint in Prometheus format.

Reviewer: I think we can drop `being` here

...
I1107 15:13:26.820577       1 producer.go:86] Sending message: value={"producerId":"strimzi-canary-client","messageId":6718,"timestamp":1636298006820} on partition=0
I1107 15:13:28.042882       1 producer.go:100] Message sent: partition=0, offset=2252, duration=152 ms
I1107 15:13:28.042908       1 producer.go:86] Sending message: value={"producerId":"strimzi-canary-client","messageId":6719,"timestamp":1636298008042} on partition=1
I1107 15:13:28.043382       1 consumer.go:193] Message received: value={ProducerID:strimzi-canary-client, MessageID:6718, Timestamp:1636298006820}, partition=0, offset=2252, duration=223 ms
I1107 15:13:28.167955       1 producer.go:100] Message sent: partition=1, offset=2252, duration=125 ms
I1107 15:13:28.168024       1 producer.go:86] Sending message: value={"producerId":"strimzi-canary-client","messageId":6720,"timestamp":1636298008168} on partition=2
I1107 15:13:28.168741       1 consumer.go:193] Message received: value={ProducerID:strimzi-canary-client, MessageID:6719, Timestamp:1636298008042}, partition=1, offset=2252, duration=126 ms
I1107 15:13:28.421214       1 producer.go:100] Message sent: partition=2, offset=2252, duration=253 ms
...
```

> Other than the metrics endpoint that is really useful to be scraped by a Prometheus server, there is a `/status` endpoint which exposes some aggregated data from such a metrics in JSON format; it can be useful for some automation where you don't have the ability to scrape and decoding the text Prometheus format but your tool can easily understand JSON text.

Reviewer: `decoding` -> `decode`

## Allow the canary to fly in the mine

The canary tool is not integrated within the Strimzi operator (yet); it could be a future improvement on the project in order to deploy it via a `Kafka` custom resource configuration.
Right now the release related installation files provide all you need to run the canary on a Kubernetes cluster alongside your Kafka cluster.
A `Deployment` file is available with a basic configuration to connect to a Kafka cluster without TLS support and exchanging messages every 10 seconds; the other main parameter needed to be changed by the user is about the bootstrap servers address.

```yaml
env:
  - name: KAFKA_BOOTSTRAP_SERVERS
    value: my-cluster-kafka-bootstrap:9092
  - name: RECONCILE_INTERVAL_MS
    value: "10000"
  - name: TLS_ENABLED
    value: "false"
```

Together with the deployment, the installation provides a `Service` for exposing the 8080 port which is used for reaching the HTTP endpoints corresponding to metrics, status as well as liveness and readiness.

## Conclusion

Monitoring a Kafka cluster by verifying that it's behaving and working properly is really simple thanks to the canary tool.

> Even when your cluster is not heavily used, it helps to check that everything is working fine to deal with the time when the real applications will start to use it.

Reviewer 3: ```suggestion
Even when your cluster is not heavily used, it helps to check that everything is working fine for the time when the real applications start to use it.
```

> It is also useful to generate some traffic for validating A/B deployment and upgrades.

Reviewer 3: ```suggestion
It is also useful for generating traffic for validating A/B deployments and upgrades.
```

> The canary is just borned with its first [release](https://github.com/strimzi/strimzi-canary/releases), so it is important for us to hear from the community about questions, issues but mostly improvements with new ideas!

Reviewer: `borned` -> `born`. I wonder if we should drop the first sentence as this will need updating as soon as we make another release.

Author: Well as any blog post it makes sense at the date you are writing it.

> The canary is just borned with its first [release](https://github.com/strimzi/strimzi-canary/releases), so it is important for us to hear from the community about questions, issues but mostly improvements with new ideas!

Reviewer 3: ```suggestion
The canary has just hatched with its first [release](https://github.com/strimzi/strimzi-canary/releases), so it is important for us to hear from the community about questions, issues, and any suggestion for improvements!
```
